fix #17630: Implement cycle detection for recursive concepts

elijahr opened the pull request:
fixes #17630

## Recursive Concept Cycle Detection

- Track (conceptId, typeId) pairs during matching to detect cycles
- Changed marker from IntSet to HashSet[ConceptTypePair]
- Removed unused depthCount field
- Added recursive concepts documentation to manual
- Added tests for recursive concepts, distinct chains, and co-dependent concepts

## Fix Flaky `tasyncclosestall` Test

The macOS ARM64 CI jobs were failing due to a flaky async socket test (unrelated to concepts).

The test only accepted `EBADF` as a valid error code when closing a socket with pending writes. However, depending on timing, the kernel may report `ECONNRESET` or `EPIPE` instead:

- **EBADF**: Socket was closed locally before kernel detected remote state
- **ECONNRESET**: Remote peer sent RST packet (detected first)  
- **EPIPE**: Socket is no longer connected (broken pipe)

All three are valid disconnection errors. The fix accepts any of them, making the test reliable across platforms.

elijahr committed:
Implement cycle detection for recursive concepts

- Track (conceptId, typeId) pairs during matching to detect cycles
- Changed marker from IntSet to HashSet[ConceptTypePair]
- When cycle detected, return true (coinductive semantics)
- Handle self-referential concepts (concept references itself in proc signature)
- Removed unused depthCount field
- Added recursive concepts documentation to manual
- Added tests for recursive concepts, deep distinct chains, and co-dependent concepts

elijahr committed:
Merge branch 'devel' of github.com:nim-lang/Nim into concepts-cycle-detection

elijahr committed:
Fix flaky tasyncclosestall test on macOS ARM64

The test was failing intermittently on macOS ARM64 because it only
accepted EBADF as a valid error code when closing a socket with pending
writes. However, depending on timing, the kernel may report ECONNRESET
or EPIPE instead:

- EBADF: Socket was closed locally before kernel detected remote state
- ECONNRESET: Remote peer sent RST packet (detected first)
- EPIPE: Socket is no longer connected (broken pipe)

All three are valid disconnection errors for this test scenario. The
test's purpose is to verify async operations don't stall after socket
close, which any of these errors confirms.

elijahr committed:
Handle early ECONNRESET during send loop in tasyncclosestall

On macOS ARM64, the kernel may return ECONNRESET immediately when
detecting a non-responsive connection, rather than letting the send
stall and timeout. This happens at line 26 during the withTimeout
call, before we ever close the socket.

The previous fix only handled errors after socket close. This fix
wraps the send/timeout in try-except to catch early disconnection
errors during the normal send loop.

Also refactored error checking into isExpectedDisconnectionError
helper to reduce duplication.

Araq commented on compiler/concepts.nim:
```diff
@@ -188,32 +190,48 @@ iterator traverseTyOr(t: PType): PType {. closure .}=
+  # one we're currently matching and has no body (n=nil).
+  if concpt.n.isNil:
+    if concpt.sym != nil and m.concpt.sym != nil and
+       concpt.sym.name.s == m.concpt.sym.name.s:
```
```suggestion
       concpt.sym == m.concpt.sym:
```

Araq commented on compiler/concepts.nim:
```diff
@@ -73,18 +73,20 @@ type
     mfCheckGeneric  # formal <- formal comparison as opposed to formal <- operand
-  
+
+  ConceptTypePair = tuple[conceptId, typeId: int]
```
```suggestion
  ConceptTypePair = tuple[conceptId, typeId: ItemId]
```

elijahr committed:
Update compiler/concepts.nim

Co-authored-by: Andreas Rumpf <araq4k@proton.me>

elijahr committed:
Update compiler/concepts.nim

Co-authored-by: Andreas Rumpf <araq4k@proton.me>

elijahr committed:
Merge branch 'devel' into concepts-cycle-detection

elijahr committed:
Merge branch 'devel' of github.com:nim-lang/Nim into concepts-cycle-detection

elijahr committed:
Merge branch 'concepts-cycle-detection' of github.com:elijahr/Nim into concepts-cycle-detection

elijahr committed:
Fix type mismatch in ConceptTypePair construction

Use itemId (ItemId type) instead of id (int type) when constructing
the ConceptTypePair for cycle detection. The tuple expects ItemId
values, not plain integers.
